move Pkg to stdlib

KristofferC opened the pull request:
Pkg3 will replace Pkg(2) but moving it to the stdlib first will make that transition easier.

On the few places Pkg is used in base, I temporarily just used a Ref that Pkg fills.

KristofferC committed:
move Pkg to stdlib

vchuravy commented on base/interactiveutil.jl:
```diff
@@ -340,11 +340,11 @@ function versioninfo(io::IO=STDOUT; verbose::Bool=false, packages::Bool=false)
+        println(io, "  Package Directory: ", PKG_MODULE_REF[].dir())
         print(io, "  Package Status:")
-        if isdir(Pkg.dir())
+        if isdir(PKG_MODULE_REF[].dir())
```
I think you might need to use `invokelatest` here. @vtjnash should be able to confirm that.

KristofferC replied:
Seems to work ok actually.

vchuravy commented on stdlib/Pkg/src/Pkg.jl:
```diff
@@ -13,9 +13,9 @@ Please see the manual section on packages for more information.
-export Dir, Types, Reqs, Cache, Read, Query, Resolve, Write, Entry
-export dir, init, rm, add, available, installed, status, clone, checkout,
-       update, resolve, test, build, free, pin, PkgError, setprotocol!
+#export Dir, Types, Reqs, Cache, Read, Query, Resolve, Write, Entry
```
These should probably be kept exported?

KristofferC commented:
So this made the Pkg tests start writing their info messages to STDOUT for some reason...

Also, the error message extremely unhelpful:

```
Pkg: Error During Test at nothing:1
  Test threw an exception of type RemoteException
  Expression: Pkg
  Error deserializing a remote exception from worker 2
  Remote(original) exception of type TestSetException
  Remote stacktrace : 
  finish at /usr/home/julia/worker/11rel-amd64/build/usr/share/julia/site/v0.7/Test/src/Test.jl:784
  macro expansion at ./util.jl:310 [inlined]
  top-level scope at ./<missing>:19
  eval at ./boot.jl:295
  #runtests#1 at /usr/home/julia/worker/11rel-amd64/build/test/testdefs.jl:25
  #runtests at ./<missing>:0 [inlined] (repeats 2 times)
  #125 at /usr/home/julia/worker/11rel-amd64/build/usr/share/julia/site/v0.7/Distributed/src/process_messages.jl:269
  run_work_thunk at /usr/home/julia/worker/11rel-amd64/build/usr/share/julia/site/v0.7/Distributed/src/process_messages.jl:56
  macro expansion at /usr/home/julia/worker/11rel-amd64/build/usr/share/julia/site/v0.7/Distributed/src/process_messages.jl:269 [inlined]
  #124 at ./event.jl:92
```

KristofferC commented:
I can't reproduce the 1 test error locally and the error message says absolutely nothing.

JeffBezanson commented:
Looks like freebsd is still trying to run `pkg` tests?

KristofferC commented:
Yeah, perhaps that can be removed @iblis?

Does anyone know the reason for splitting the tests into two parts for CI? Something with online / offline tests but `choosetests` already checks Internet connectivity and filters based on that.

iblis17 commented on .travis.yml:
```diff
@@ -144,7 +144,7 @@ script:
     - pushd /tmp/julia/share/julia/test
     - /tmp/julia/bin/julia --check-bounds=yes runtests.jl $TESTSTORUN &&
-      /tmp/julia/bin/julia --check-bounds=yes runtests.jl libgit2-online download pkg
+      /tmp/julia/bin/julia --check-bounds=yes runtests.jl libgit2-online download
```
`.circleci/config.yml` need to be changed as well?

KristofferC replied:
Yes, I didn't know that existed, thanks.

iblis17 commented:
I think I should provde cumstom shell script like Travis and CircleCI does in long-term...
(Let me figure out how to make it

iblislin committed:
introduce customizable build scripts for FreeBSD CI

See also:
  - https://github.com/JuliaLang/julia/pull/25705#issuecomment-360701168
  - https://github.com/JuliaLang/julia/pull/25705#issuecomment-360961827

[av skip]

iblislin committed:
single script

KristofferC commented:
All CI now fail with 

```
julia: /home/travis/build/JuliaLang/julia/src/codegen.cpp:1096:
jl_llvm_functions_t jl_compile_linfo(jl_method_instance_t**, jl_code_info_t*, size_t, const jl_cgparams_t*):
Assertion `li->min_world <= world && li->max_world >= world' failed.
```

During precompile. Perhaps adding some debug printing to `precompile.jl` would be useful?

StefanKarpinski commented:
cc @vtjnash ^

KristofferC committed:
comment out world age assertion

vchuravy commented:
I would suspect that the worldage error is: https://github.com/JuliaLang/julia/pull/25705#discussion_r163283315

iblis17 commented on .freebsdci.sh:
```diff
@@ -0,0 +1,59 @@
+
+    gmake testall \
+        test-download \
+        test-Pkg \ # <- TODO: Should only test Pkg/pkg but no rule for test-Pkg/pkg
```
The comment in this line neutralized the effect of `\` , so FreeBSD CI build failed.

line 73:
https://freebsdci.julialang.org/#/builders/1/builds/7031/steps/7/logs/stdio

KristofferC replied:
Oops, thanks.

KristofferC replied:
Is there anyway we could run this as `julia test/runtests.jl test-Pkg/pkg`?

iblis17 replied:
I think invoking `./usr/bin/julia` should work

KristofferC commented:
@vchuravy But the `PKG_MODULE_REF` thing doesn't exist at all anymore on the newer commits because enough stuff got moved to stdlib that it wasn't needed anymore.

KristofferC commented on .freebsdci.sh:
```diff
@@ -36,10 +36,8 @@ runtests(){
-        test-download \
-        test-pkg \
-        test-libgit2-online
+    ./usr/bin/julia test all
```
@iblis17 Did this mess up for the builder?

iblis17 commented on .freebsdci.sh:
```diff
@@ -36,10 +36,8 @@ runtests(){
-        test-pkg \
-        test-libgit2-online
+    ./usr/bin/julia test/runtests.jl all
+    ./usr/bin/julia test/runtests.jl libgit2-online Pkg/pkg download
```
need `--check-bounds=yes`?

KristofferC replied:
Don't think so, already gets applied at https://github.com/JuliaLang/julia/blob/master/test/testenv.jl#L20

iblis17 replied:
👌

KristofferC commented:
The world error assertion was disabled from consulting `@vtjnash` and he said it wasn't really a problem disabling it and that he would look at it later.

StefanKarpinski commented:
![giphy-1](https://user-images.githubusercontent.com/153596/35576652-885bbb6e-05ad-11e8-8ba4-311b816e51c0.gif)
